### PR TITLE
Consolidate keyboard shortcuts into a single registry

### PR DESCRIPTION
**Pull request by donkawechico**
## Proposed change

Now that other folks are starting to introduce shortcuts into Home Assistant (https://github.com/home-assistant/frontend/pull/7207, https://github.com/home-assistant/frontend/pull/7230), it might be a good idea to consolidate that logic into a single place and a standardized API.

Eventually this could be used to both a) warn about, or ensure we don't overwrite existing hotkeys, and b) surface all existing application hotkeys to the user (say, in a helpful shortcut overview launched with the question mark).

This doesn't (yet) have any protections against folks overwriting shortcuts by accident, but it does make it easier to see what components are registering hotkeys, and which keys are already taken.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/frontend/issues/7327
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/scene/ha-scene-editor.ts [hunk 1]
```diff
@@ -414,6 +417,10 @@ export class HaSceneEditor extends SubscribeMixin(
     `;
   }
 
+  protected firstUpdated() {
+    this.registerShortcut(HAKeyboardShortcut.CTRL_S, () => this._saveScene());
+  }
```
Comment (donkawechico) [inline src/panels/config/scene/ha-scene-editor.ts:422]: This works, but it breaks if I use the standard shorthand for void callbacks:

```
protected firstUpdated() {
    this.registerShortcut(HAKeyboardShortcut.CTRL_S, this._saveScene); // <-- saveScene fails from corrupted "this"
}
```

By the time we get into `_saveScene` method, the `this` has been corrupted and doesn't have the needed properties. 

If anyone can help explain why I'd be happy to find a solution.

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 2]
```diff
@@ -1,26 +1,64 @@
...
-    };
+    private _registeredShortcuts: RegisteredShortcuts = {};
 
-    public connectedCallback() {
-      super.connectedCallback();
-      this.addEventListener("keydown", this._keydownEvent);
+    protected registerShortcut(
+      shortcut: HAKeyboardShortcut,
+      callback: () => void,
+      element: LitElement | Document = this
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:23]:
```suggestion
      element: HTMLElement | Document = this
```

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 3]
```diff
@@ -1,26 +1,64 @@
...
+            }
+            break;
+        }
+      }
+    };
+
+    private isOSCtrlKey(e: KeyboardEvent) {
+      const isMac = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
+      return isMac ? e.metaKey : e.ctrlKey;
+    }
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:63]: This check should be done once.

File: src/panels/config/scene/ha-scene-editor.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/config/scene/ha-scene-editor.ts:422]: Your current code is fine. Arrow functions keep the context, normal functions create a new context.

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 3 above)
Comment (donkawechico) [reply src/mixins/keyboard-shortcut-mixin.ts]: Done

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 2 above)
Comment (donkawechico) [reply src/mixins/keyboard-shortcut-mixin.ts]: Done

File: .eslintrc.json [hunk 4]
```diff
@@ -75,11 +75,13 @@
     "object-curly-newline": 0,
     "default-case": 0,
     "wc/no-self-class": 0,
+    "no-shadow": 0,
     "@typescript-eslint/camelcase": 0,
     "@typescript-eslint/ban-ts-comment": 0,
     "@typescript-eslint/no-use-before-define": 0,
     "@typescript-eslint/no-non-null-assertion": 0,
     "@typescript-eslint/no-explicit-any": 0,
+    "@typescript-eslint/no-shadow": ["error"],
```
Comment (donkawechico) [inline .eslintrc.json:84]: Needed due to issue with making `enum` types in typescript with default eslint.

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 5]
```diff
@@ -1,26 +1,65 @@
...
+        const shortcutAction = this._registeredShortcuts[shortcut];
+        if (shortcutAction) {
+          shortcutAction();
+        }
+      }
     }
 
-    protected handleKeyboardSave() {}
+    private _keydownEvent = (event: Event) => {
+      const _event = event as KeyboardEvent;
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:45]: Why not?
```suggestion
    private _keydownEvent = (event: KeyboardEvent) => {
```

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 6]
```diff
@@ -1,26 +1,65 @@
...
+      shortcut: HAKeyboardShortcut,
+      callback: () => void,
+      element: HTMLElement | Document = this
+    ) {
+      element.addEventListener("keydown", this._keydownEvent);
+      this._registeredShortcuts[shortcut] = callback;
     }
 
-    public disconnectedCallback() {
-      this.removeEventListener("keydown", this._keydownEvent);
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:21]: Why did we remove this?

**Comment by KevinCathcart** [comment]
One question to consider is the approach for supporting keybindings. The approach shown here works, but if we get to 10-20 shortcuts, the pattern used here will begin to expand to quite a bit of code (the switch style decoding of the keypresses in particular).

Might it be better to make use of a more traditional approach to avoid code bloat? Perhaps it makes sense to use a small keybinding library like the ~400B when minified [tinykeys library](https://github.com/jamiebuilds/tinykeys), or borrow some code from such a library.

Separately, it might be better to have shortcut actions registered without a specific keybinding specified in the registering code. Instead have the registry itself have a static mapping of actions to bindings. That approach would make it rather straightforward to eventually add customizable keybindings, if demand should arise (the existing map becomes the defaults).

I'm suggesting the above only as things that might be worth thinking about. I am not objecting to the current proposed code, since it is unquestionably an improvement over the current situation.

**Comment by bramkragten** [comment]
tinykeys does look nice :-)

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 5 above)
Comment (donkawechico) [reply src/mixins/keyboard-shortcut-mixin.ts:45]: I uh... er... um... well you see...

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 6 above)
Comment (donkawechico) [reply src/mixins/keyboard-shortcut-mixin.ts:21]: Not sure. I probably just saw that `connectedCallback` wasn't needed anymore so I reflexively remove disconnected as well.

I put it back.

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 5 above)
Comment (donkawechico) [reply src/mixins/keyboard-shortcut-mixin.ts]: Done

**Comment by KevinCathcart** [comment]
Yeah, I was certainly trying to suggest that if it were used, it would be used as part of the implementation of the mixin. Having every-component set up shortcuts themselves with no coordination is exactly what we don't want.

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 5 above)
Comment (KevinCathcart) [reply src/mixins/keyboard-shortcut-mixin.ts]: That is strange. If using even a remotely recent version of typescript (i.e. from some point in 2018 onwards) and the first parameter is "keydown"/"keyup"/"keypress", it is supposed to enable an overload with a callback that takes `KeyboardEvent`. You can find the [mapping here](https://github.com/microsoft/TypeScript/blob/7a484695ca3452b29b268d037ce1d7dda9d69761/lib/lib.dom.d.ts#L5717). Warning: Positively enormous file (20k lines long), which could potentially crash a browser.

Wonder if there is some form of overload resolution bug being hit here. Oh well, that workaround is reasonable enough.

I've seen [some indication](https://github.com/jamiebuilds/tinykeys/pull/12) that at least one browser or js framework at one point or other would pass events that were not derived from KeyboardEvent (and not fully duck type compatible) in some edge cases, but the typescript type definitions definitely do not that that into account. I have no positive knowledge of this occurring, so this is just hearsay.

**Comment by bramkragten** [comment]
Still, this would not be able to tell what elements use what keyboard shortcuts, we would need a global registry for that. This mixin will only know the keys that the current element registered.

**Comment by donkawechico** [comment]
>Still, this would not be able to tell what elements use what keyboard shortcuts, we would need a global registry for that. This mixin will only know the keys that the current element registered.

Agreed. As I mention in the description, having the logic consolidated into one place should passively help avoid shortcut clobbering for now. But eventually I agree we need a more global dictionary.

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 7]
```diff
@@ -1,26 +1,69 @@
...
-    };
-
-    public connectedCallback() {
-      super.connectedCallback();
-      this.addEventListener("keydown", this._keydownEvent);
-    }
+    private _registeredShortcuts: RegisteredShortcuts = {};
 
     public disconnectedCallback() {
       this.removeEventListener("keydown", this._keydownEvent);
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:23]: We remove the listener on disconnected, we should add it again on connected. The firstupdated will only run once, so after the element is disconnected the listener will never be added again on connect

File: src/mixins/keyboard-shortcut-mixin.ts [hunk 8]
```diff
@@ -1,26 +1,69 @@
...
+        }
+      }
+    }
+
+    protected registerShortcut(
+      shortcut: HAKeyboardShortcut,
+      callback: () => void,
+      element: HTMLElement | Document = this
+    ) {
+      element.addEventListener("keydown", (event) => this._keydownEvent(event));
```
Comment (bramkragten) [inline src/mixins/keyboard-shortcut-mixin.ts:45]: We should not do this in `registerShortcut`, we would get multiple listeners if you register multiple shortcuts.

File: src/mixins/keyboard-shortcut-mixin.ts (hunk 7 above)
Comment (bramkragten) [reply src/mixins/keyboard-shortcut-mixin.ts:23]: Also, this will not remove it now, as the callbacks for the add and remove are different.

**Comment by bramkragten** [comment]
Do we need to update this, or just close?

**Pull request closed without merging**